[Pun Hui Min] iP

punnyhuimin opened the pull request:
# _Ducky_ 🦆 
> "Be like a duck. Calm on the surface, but always paddling like the dickens underneath." - Micheal Caine ([Source](https://quotecatalog.com/quote/michael-caine-be-like-a-duck-baVAqop#:~:text=%E2%80%9CBe%20like%20a%20duck.,paddling%20like%20the%20dickens%20underneath.%E2%80%9D))

### Ducky is an interactive tool for you to track your tasks in one place. It's
- Cute ✧･ﾟ: *✧･ﾟ:* |\__( o)> *:･ﾟ✧*:･ﾟ✧
- ~~Easy~~ **Very _easy_** to use
- Text-based 🧑‍💻 

### All you need to do is, 
1. Download it [here](https://github.com/punnyhuimin/ip/releases/tag/A-Jar) 👈 
2. Double click it 🖱️ 🖱️ 
3. Add your tasks 📆 
4. Let it manage your tasks for you! (๑•̀ㅁ•́๑)✧

## Features: 
- [X] Welcomes you to the CLI 🔆 
- [X] Says goodbye 👋 
- [X] Provides help when needed (Just type `help`) 🙇 
- [X] Managing events with dates
- [X] Managing todos
- [X] Managing deadlines with due dates
- [ ] Reminders (Coming soon)

If you are a Java programmer, you can use it to practice Java too! Here's a small snippet of the `run` code: ☘️ 

```java
public void run() {
    ui.welcome();
    boolean isExit = false;
    while (!isExit) {
        try {
            String fullCommand = ui.readCommand();
            // show the divider line ("_______")
            ui.showLine();
            Command c = new Parser(fullCommand).parse();
            c.execute(tasks, ui, storage);
            storage.saveFile(tasks.formatTasks());
            isExit = c.isExit();
        } catch (DukeException e) {
            ui.showError(e.getMessage());
        } finally {
            ui.showLine();
        }
    }
}

```

Commit:
Add Gradle support

punnyhuimin committed:
Updated Duke class

Duke class now accepts user inputs and returns the same input to the
user. When input is "bye", the class will return a different message.

punnyhuimin committed:
Updated Duke with list handler

Ducky, previously known as Duke, can now store the text handed to it and
display the text back to the user when they request for it.

To enhance Ducky, next step is to add the ability to mark tasks as done.

punnyhuimin committed:
Added Task class and updated Duke class

Added ducky logo to beautify output.

New Task class was created for handling each task created by the user.

User can now mark their tasks as done, or mark as not done.
Instructions are also given when the user types the "help" command.

Next step to improve the application is to set up the ability for user
to add task type i.e. event, todo, deadline.

punnyhuimin committed:
Created Deadline, Todo, Event and Updated Duke

Updated 'help' description to include new features that the text that
user can pass into Ducky. Created Deadline, Todo, Event classes that
inherit from the Task class to support task tracking.

Extracted common behavior from the Duke class output to reduce code
duplication. i.e. the long line printed after each run.

Next step is to set up semi-automated testing of Duke class.

punnyhuimin committed:
Edited EXPECTED.TXT

Having issues with the Ducky logo, not able to pass the automated tests.
Might be due to spacing error as the test passes without the logo.

More trials to be conducted to ensure Ducky logo can be included

punnyhuimin committed:
Added input.txt file and updated EXPECTED.TXT

Updated Ducky logo in Duke.java to remove whitespaces. The whitespaces
were not being caught in the EXPECTED.TXT and should not have been there
in the first place.

Added test cases for all possible inputs (e.g. help, deadline, todo,
event, bye, list, mark, unmark)

punnyhuimin committed:
Refactored Duke

Duke had a lot of if else statements being used in the main file. I
split the sections into smaller chunks to ensure that the code is
readable and can be easily cleaned up moving forward. This will also aid
in the exception handling later.

punnyhuimin committed:
Added DukeException, Updated Duke, and Added test cases

Added documentation to Duke to showcase the params the methods take and
what it outputs.Exceptions handled include wrong date input format, no
description, no date provided, no proper instructions, and calling an
empty list.

Added more test cases to check on whether he correct exceptions will be
thrown.

Date validity should be checked in each Event and Deadline. Error
should be thrown there to keep Duke code to a minimum.

punnyhuimin committed:
Updated Duke with delete functionality and Updated EXPECTED.txt

Added delete functionality and exception handling for the functionality.
This includes handling of no index provided and inaccurate index
entered.

Updated the test files to test out the delete functionalities and all
error handling methods are working.

punnyhuimin committed:
Added JavaDocs to Duke, DukeException, and Task

JavaDocs are added to all methods within Duke, DukeException, and Task.

Will need to add in more JavaDocs as the code progresses. May want to
also add it for Event, Deadline, and ToDo classes.

punnyhuimin committed:
Created Checker class

Checker class now handles the enum, allowing for switch case use in Duke
to keep the code as tidy as possible. All other functions, such as the
addition of deadlines or todos, still remain the same.

Added a new string to set as the String to be returned when no
description is to be found. Provides ease of change for the future.

JavaDocs needs to be done for the Checker class.

punnyhuimin committed:
Implemented task saving to duke.txt

Created a new TextWriter class which handles writing of tasks into a
text file. Added exception handling to Textwriter as well.

New Exception handled in the case where the command Deadline does not
come with "/by" or Event does not come with "/at.

Added two new test cases to check for the new Exception being handled
(i.e. No /at or /by command provided).

Will need to write JavaDocs for the TextWriter class.

punnyhuimin committed:
Added date functionality to Deadline and Event

Created a new class "DateHelper" that helps with converting strings to
Date class. DateHelper stores the dates and can format the date into
the required formats to be outputted in Deadline and Event.

Updated input.txt and EXPECTED.TXT to include the new date formats.

JavaDocs to be updated for DateHelper, TextWriter, and Duke.

punnyhuimin committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Implemented task saving to duke.txt
  Created Checker class

punnyhuimin committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Added date functionality to Deadline and Event

punnyhuimin committed:
Completed OOP of Parser, Storage, TaskList, and Storage

Edited all files to help keep Duke as lean as possible. Parser is an
enum class that helps to check which command was given. Storage helps
read and write files. Every change in the list will be marked down in
the textfile immediately. Time is also handled in a DateHelper.

JavaDocs to be updated and do error handling for mark and unmark. Mark
and unamrks' error in id passed is not properly handled.

punnyhuimin committed:
Added Command class and Packages

Added Command class which is an abstract class. This class is made to
handle the different commands that could be executed by Ducky. Classes
include:
- AddEventCommand
- AddDeadlineCommand
- AddTodoCommand
- MarkCommand
- UnmarkCommand
- ExitCommand
- ListTaskCommand
- HelpCommand
- DeleteCommand

Packages were also created to add in classes of the same types. Three
packages created were:
- Duke
- Command
- Task
Command and Task are packages under the Duke package.

JavaDocs needs to be completed further.

punnyhuimin committed:
Created JUnit Tests for Ui and Checker class

Tested the Checker class to see if it is returning the correct enum
based on the command given into it. Tested UI class methods like
welcome, showLine, showError, and printHelp.

More tests to be created for the other classes.

punnyhuimin committed:
Added JavaDocs and corrected issue in AddEventCommand

AddEventCommand was adding Deadline task instead of Event task, causing
the list generated to be incorrect. Error has since been corrected.

JavaDocs have been added to the code to provide creators with
understanding of each methods' usage.

punnyhuimin committed:
Formatted parts of code to follow coding standards

Formatted comments to be on the top of what it explains instead of
being by the side.

punnyhuimin committed:
Added Find functionality

Duke can now find specific tasks that contains the words specified by
the user. It can help the user to quickly search for the task that
concerns them at this point in time.

punnyhuimin committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Formatted parts of code to follow coding standards

# Conflicts:
#	src/main/java/duke/Storage.java

punnyhuimin committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Added Find functionality

# Conflicts:
#	src/main/java/duke/Checker.java
#	src/main/java/duke/task/TaskList.java

ChanWeiJie commented on src/main/java/duke/Checker.java:
```diff
@@ -0,0 +1,68 @@
+     *
+     * @return State based on the input. E.g. help, delete, event.
+     */
+    public Status getStatus(){
```
With regards to the coding standard, there should be a space after <code> getStatus</code>, I noticed the same issue in several other places too. But other than that good job ! 😄

ChanWeiJie commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,111 @@
+        String[] parts = line.split("\\|");
+        String state = parts[0].trim();
+        String type = parts[1].trim();
+        String description = parts[2].trim();
```
Perhaps declaring a few variables such as 
<code>private final String DESCRIPTION_OFFSET = 2</code> or  
<code>private final String TIME_OFFSET = 3</code> for the "magic numbers" 2/3 used here ?

ChanWeiJie commented on src/main/java/duke/command/AddEventCommand.java:
```diff
@@ -0,0 +1,57 @@
+
+    private static final String DESC_RESPONSE = "Oops! \\(@.@)/ You have not keyed in a description for the task!\n" +
+            "Let's try again ~(^.^)~\n" +
+            "Type 'help' if you need to know how to use this command";
```
With regards to the coding standard, the break should be before an operator.  I noticed the same issue in several other places too. 
Perhaps such as :
 <code> "Oops! \\(@.@)/ You have not keyed in a description for the task!\n" </code>
         <code>     + "Let's try again (^.^)\n" </code>
         <code>   +  "Type 'help' if you need to know how to use this command"; </code> ? 🤔

ChanWeiJie commented on src/test/java/CheckerTest.java:
```diff
@@ -0,0 +1,65 @@
+public class CheckerTest {
+
+    @Test
+    public void checkerTest() throws DukeException {
```
I like how you extracted all the exception checking into one class instead of the checking them 1 by 1 in the parser class. This also makes this bit of code clean! 👍

ChanWeiJie reviewed (commented):
LGTM! Just some nix to fix.

punnyhuimin committed:
Reformatted code to align with coding standards

Missed out certain formatting areas e.g. spacing after function before
curly brackets. Formatting is now set to the standard required.

punnyhuimin committed:
Merge branch 'add-gradle-support' of https://github.com/punnyhuimin/ip

* 'add-gradle-support' of https://github.com/punnyhuimin/ip:
  Add Gradle support

punnyhuimin committed:
Added Gradle to automate project builds

Gradle is now set up and can build and run Duke. To run it, go to
Tasks > application > right click 'run' and run it. This will start the
Ducky CLI.

Gradle can also run JUnit tests. There was an error in the
`./gradlew clean test` initially, where a class could not be
initialized. Error was resolved by changing the Gradle version to 6.4.1.

punnyhuimin committed:
Added checkstyle.xml and suppressions.xml

These files will help with checkstyle. IntelliJ can now check the style
of the Duke code to see if it is compliant with the code formatting
required.

punnyhuimin committed:
Merge branch 'branch-A-Gradle'

* branch-A-Gradle:
  Added Gradle to automate project builds

punnyhuimin committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Added checkstyle.xml and suppressions.xml

alfredkohhh reviewed (commented):
Looks good to me! I like how you separated the specific "bye" input, and gave a different message

alfredkohhh commented on src/main/java/Duke.java:
```diff
@@ -1,12 +1,18 @@
 import java.io.*;
```
Good usage of individual imports! For this particular import for io, could this be extracted out to be individual imports as well ?

alfredkohhh reviewed (commented):
Look awesome so far! Just a few nix to fix

punnyhuimin committed:
Created GUI for Ducky

Ducky now runs on a GUI. All the commands are the same as what it was in
the CLI.

Improvements to be done include:
1. Ducky should welcome the user.
2. Ducky image should be rounded and smaller to give space to the text.
3. User image should be changed to something nicer.
4. Current DialogBox is getting smaller when there are more texts. Issue
can be replicated by typing "help" at every step.

punnyhuimin committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Created GUI for Ducky

punnyhuimin committed:
Updated GUI for better UI

GUI now has the ducky image. User image is also changed to an egg,
which fits with the yellow theme of the application.

Caught errors with the Parser class as it was not catching the case
where the user only mentions the command and nothing else. This was
causing the GUI to become buggy and resolving it was of top priority.

Assertions should be used alongside Exceptions in the classes to further
ensure that there will be no uncaught errors moving forward.

punnyhuimin committed:
Added assertions

Added assertions to all Command classes and to the TaskList class. The
addition of assertion helps to ensure that the different tasks being
added are not empty. It also ensures that the description of the task is
not empty when passed into the Command classes for task creation. It
also helps to indicate a possible bug in the code as it violates an
assumption about how the code should be performed.

More assertions can be included in the Task classes to ensure that the
created classes have the necessary parameters to perform as expected.

punnyhuimin committed:
Improved code quality

Removed magic numbers and strings to improve readability of the code.
Replaced certain if else conditions with booleans and named them based
on what they are supposed to show. e.g. isEmptyList.

Removed run method in Duke as it was no longer being used. More can be
done to improve the code quality. Some files have not been vetted for
magic numbers and strings

punnyhuimin committed:
Merge pull request #2 from punnyhuimin/branch-A-CodeQuality

Improved code quality

punnyhuimin committed:
Merge branch 'master' of https://github.com/punnyhuimin/ip

* 'master' of https://github.com/punnyhuimin/ip:
  Improved code quality

punnyhuimin committed:
Merge branch 'master' into branch-A-Assertions

punnyhuimin committed:
Merge pull request #1 from punnyhuimin/branch-A-Assertions

Added assertions

punnyhuimin committed:
Merge branch 'master' of https://github.com/punnyhuimin/ip

* 'master' of https://github.com/punnyhuimin/ip:
  Added assertions

punnyhuimin committed:
Added natural date functionality

Ducky can now detect inputs such as Monday and Mon to set the due dates
and event dates as the next nearest Monday / day provided.

punnyhuimin committed:
Add functionality for exit command

Upon bye command, the dialog closes in 3 seconds.

punnyhuimin committed:
Highlight errors in user command

When Ducky receives a command that is not part of its intended use, the
dialog box turns bright orange to signal that the user has inputted
something wrongly.

punnyhuimin committed:
Change Sad Ducky image

Change SadDucky image to royalty free image.

punnyhuimin committed:
Clean up unused commands and Javadocs

Cleaned up unused commands such as 'getResponse' and added more JavaDocs
 in Ui class. Added more functions to the help command as well.

punnyhuimin committed:
Edit find command and duke error handling

Ensured that duke's error variable is not accessed directly from
MainWindow. Created a new function to allow MainWindow to access the
error boolean instead.

Find command was faulty and required slight revamp to account for a
missing newline.

punnyhuimin committed:
Set theme jekyll-theme-minimal

punnyhuimin committed:
Add user guide

User guide now details all functionalities that Ducky provides, and
gives use cases for them as well.

punnyhuimin committed:
Merge commit '85c5e52f83880b2c7a70f06c5ad15c5b5aa8ff1c'

* commit '85c5e52f83880b2c7a70f06c5ad15c5b5aa8ff1c':
  Set theme jekyll-theme-minimal

punnyhuimin committed:
Edit user guide

Formatting issues with user guide now accounted for. Added ui image.

punnyhuimin committed:
Add background to GUI and update Ui image

GUI of Ducky now has a garden background to fit Ducky's habitat. The Ui
image has also been updated accordingly.

punnyhuimin committed:
Add new Ui image

New Ui image now showcases some commands used. Ui is shown in the user
guide as well.

punnyhuimin committed:
Update README.md

More changes to be made regarding things to expect upon code launch. Might want to consider downloading the JAR file instead of code.

punnyhuimin committed:
Update README.md

Added UI Image for better preview of the project.

punnyhuimin committed:
Update README.md

Changed it to user guide information. Updated developer setup information as well.

punnyhuimin committed:
Update resource usage in DateChecker

punnyhuimin committed:
Update reference

punnyhuimin committed:
Update references

punnyhuimin committed:
Update spelling error
